### PR TITLE
ENH: option to reuse binnumbers in stats.binned_statistic_dd

### PR DESCRIPTION
**Pull request by egouden**
Dear scipy developers,

This is my first pull request to the scipy project !

It adds the functionality of reusing the bin numbers in scipy.stats.binned_statistic_dd.
This is useful when new values are available for the same dataset.
A quick benchmark shows a decrease in computation time by a factor 3.

Best regards,

Edouard

### CODE REVIEW

File: scipy/stats/_binned_statistic.py
Comment (tylerjereddy) [inline scipy/stats/_binned_statistic.py]: same comment here--`binnumbers` arg may need to be moved to the end if experts find this approach helpful

**Comment by rlucas7** [comment]
Thanks for your contribution @egouden 
I see a few points to suggest for change, please let me know if something is not clear.

1. I agree w/Tyler's point, generally prefer new arguments to be at the end of the argument list.
2. Add a unit test for the new feature. You can take a look at the tests here:
https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_binned_statistic.py
to get an idea of how to structure etc.
3. How should `bins=` argument work with the `binnumbers`? e.g. I can foresee a case when a user enters `bins=3, binnumbers=[1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0]` so the lengths are not compatible. Function should throw an error in this scenario. 
4. I would recommend to ***add*** an additional example in the docstring that provides example usage. 
5. Should the argument name be `binnumbers` or `binedges`? 
I'm not entirely clear on the usage yet, likely with a docstring example and a test case I will be more clear.

**Comment by egouden** [comment]
Thank you for your comments @rlucas7 
Here are my answers:

1. Indeed the new argument has to be put at the end for backward compatibility. This kind of limitation could be solved in the future by using @[Keyword-Only Arguments](https://www.python.org/dev/peps/pep-3102/).
2. A test has already been added.
3. The bin numbers are the result of the sorting of the sample points in the bins. It is obtained by a previous call to the function. This is now clearer in the documentation. The bins argument had to be consistent indeed which was tedious. But now it is only based on the precomputed bin edges.
4. I added a doc test.
5. See answer to 3.

To implement the changes I had to increase the modularity of the code.

**Comment by rlucas7** [comment]
Thanks for the changes @egouden 
I'm pretty busy until the upcoming weekend. 
I'll block a some time on either Saturday or Sunday (Sept 27th or 28th) to take a closer look, run the tests locallly on my machine etc. 
So poke me if you don't see any follow up comments by Sunday evening :).

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: This has been solved.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: The named tuple expects to contain names, `statistic`, `bin_edges`, and `binnumber` these are guaranteed if the `reuse` value is a `binnedStatisticddResult` but not if the namedtuple is one the user came up with on their own. Probably want to make that explicit in the docstring. The next line here underneath says `previous call to the function` in which case the input ***must*** be a `binnedStatisticddResult`.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: nit: add a trailing comma e.g. `'max']` -> `'max',]`

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: the if/else block above has the same logic e.g. `if resue is None:` etc. why can't all the logic be in 1 if/else block rather than 2?

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: I notice now that `reuse` uses the `bin_edges` and the `binnumber` elements in the named tuple but not the `statistic` the usage should be spelled out in the docstring. e.g. some users may expect to be able to leave the `statistic` argument to the `binned_statistc_dd()` call empty in which case they might get an error. Also it could be that the two disagree (e.g. named tuple uses `count` but in the function signature the `statistic=std`. Which statistic should the user expect to be returned?

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: this commented line is an atavism from before `_bin_edges` was factored out. The comment doesn't seem to hold anymore, is that correct behavior?

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: even though the function is 'private' there should still be a docstring, it makes maintenance and future debugging a lot easier.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: this e.g. converts `int` input types into `float`. I realize this was behavior from before but if the float part is dropped here I think we would retain the type in the numpy array

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: just a comment, here it makes sense to keep the `edges` as float.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: what if `dedges[i].min()` is 0? probably want to handle that, admittedly rare, case.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: is there a reason why these two new methods are returning plain tuples? Not clear to me they need to be like this, wouldn't it be easier to have e.g. `return binnumbers`?

**Review by rlucas7 — COMMENTED** [review]
@egouden thanks for the updates to this PR. I had the chance to give it a closer look. The use case you have in mind makes more sense to me now. There is still some potential incoherence with `statistic` though (please see my CR comments for details). There are also a few other items to clarify. Getting closer though.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: This is now explicit.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: This makes flake8 unhappy.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: The two blocks have been merged

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: I guess this is solved.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: I do not understand.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: I added a short description of the private functions but I think the improved modularity already makes maintenance easier.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: I do not understand.
Feel free to change it.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: I don't know the best practices for parameter validation.
Feel free to add it.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: This is corrected.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py]: > I don't know the best practices for parameter validation.

Typical approach would be a try/except block. Throw a valueError if `dedges[i].min()` is `0.0`

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py]: hmm, ok let's scratch that then. This ***is*** valid python but we've got flake8 configured to not have that in the codebase. If we've got that set that way no sense to change for only this line.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py]: I'm asking in particular about the comment:
`# Used only if number of bins is given.`
What if number of bins is not given.

**Review by rlucas7 — COMMENTED** [review]
@egouden I'm still not clear what to do if the statistic argument to `binned_statistic_dd()` differs from the `statistic` in the namedtuple. I looked at the namedtuple results a little closer and it looks like what is being returned in `statistic` entry of the namedtuple is the statistics in each cell and not a string to indicate the name of the statistic.

The argument in the signature of the function call would be the statistic that is used. This might be unclear to some users (as it was to me) so if you could please add a sentence to the effect that would help clear up the  confusion.

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: This case has been handled and a test has been added.
But should we really handle any possible wrong inputs?

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: The comment is still relevant.
This part is actually not used if the bins are given explicitly.

**Comment by egouden** [comment]
I hope the description of the new optional arguments is more clear.
I also corrected some flake8 warnings in the tests.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py]: @egouden I agree the added modularity helps. Docstrings are helpful for when you come back to the function after several months/years and can't remember the specifics.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py]: I guess the `smin` and `smax` are being used to determine bidedges which we do want to be floats so lets leave as is.

**Comment by rlucas7** [comment]
> I hope the description of the new optional argument is more clear.
> I also corrected some flake8 warnings in the tests.

It looks like the 2 build failures are related to the valueerror test that was added.

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: change to:
```
To reuse bin edges, `expand_binnumbers` must be set to false (the default). 
```

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: NIT: the space here isn't needed

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/_binned_statistic.py]: >The comment is still relevant.
>This part is actually not used if the bins are given explicitly.

Ok, maybe I'm misunderstanding something but it sounds like the comment is still relevant but should be changes to something like `not used if bins are given explicitly`?

File: scipy/stats/_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/_binned_statistic.py]: can the message here instead read: 
`The smallest edge difference is numerically 0.`  ?

File: scipy/stats/tests/test_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/tests/test_binned_statistic.py]: why the change here?

File: scipy/stats/tests/test_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/tests/test_binned_statistic.py]: I'm thinking the name is not as informative in 2 years time. Can you add a comment line in this test, something to the effect that => 
`#NOTE: tests the reuse of bin_edges from previous call`
?

File: scipy/stats/tests/test_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/tests/test_binned_statistic.py]: thanks for adding this!

**Review by rlucas7 — CHANGES_REQUESTED** [review]
Thanks @egouden 
I left a few more comments, mostly minor things related to docstrings and unneeded comment changes. Once those changes are reflected in the PR, I'm happy to approve.

File: scipy/stats/tests/test_binned_statistic.py
Comment (egouden) [reply scipy/stats/tests/test_binned_statistic.py]: flake8

File: scipy/stats/_binned_statistic.py
Comment (egouden) [reply scipy/stats/_binned_statistic.py]: flake8

**Review by rlucas7 — APPROVED** [review]
I'm comfortable with these changes and this would be an improvement over existing implementation by adding the reuse functionality. Maybe @tylerjereddy has time and can give it a look before a merge? 

The Travis build failure looks unrelated to these changes (it is an `io` related failure).

**Comment by rgommers** [comment]
TravisCI failure is unrelated (build failure in SuperLU)

**Comment by rgommers** [comment]
This looks good to go to me too now, merging. Thanks @egouden, and thanks @rlucas7 for the thorough review!

**Comment by rgommers** [comment]
@egouden would you mind adding a release note for this new feature in https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0?

**Comment by egouden** [comment]
The release notes have been updated. I am really happy to contribute.

@rlucas7 @rgommers  @tylerjereddy Thank you for the reviews. I found the process quite long because we also improved several small things not related to the original idea, which was quite basic. I am also wondering how we could automatically integrate requested changes.

So this was my actual first PR! I have a second more ambitious PR waiting for reviews...

**Comment by rgommers** [comment]
> The release notes have been updated. I am really happy to contribute.
>
> So this was my actual first PR! I have a second more ambitious PR waiting for reviews...

Thanks @egouden! Always great to have a new contributor who enjoys getting their first PR merged and then dives into a next one:)

 > I found the process quite long because we also improved several small things not related to the original idea, which was quite basic. I am also wondering how we could automatically integrate requested changes.

Thanks for this feedback. It's very valid. Partly back-and-forth is needed, but it's possible to improve this review process. One thing that can be done and I sometimes do is to directly fix small issues in a contributor's branch. Maintainers can push changes directly to the branch a PR was submitted from (minor complication here is that @rlucas7 can't, since he doesn't yet have commit rights). This is something where a reviewer has to find a balance. At the start it's good to point out all issues. When we get close to merging or when it's as much time to point out (e.g.) docstring formatting issues as it is to just fixing them, the latter may be preferable sometimes.

Something that's harder to fix is that we're simply short on reviewer bandwidth. An all-volunteer project with ~10 million users tends to be that way .....

Let's see how your second PR goes:)

**Comment by rgommers** [comment]
By the way, this PR broke our wheel tests because it raises a divide-by-zero warning sometimes: https://travis-ci.org/MacPython/scipy-wheels/jobs/596997894. If you have time to look into it, that would be very helpful. Otherwise I'll fix it later, but haven't had time yet.

**Comment by rlucas7** [comment]
> > The release notes have been updated. I am really happy to contribute.
> > So this was my actual first PR! I have a second more ambitious PR waiting for reviews...

Thanks @egouden sorry if my review was a bit painful. I think it will be easier as you submit more PRs, at least that has been my experience. Partly from learning more about the project; e.g. goals + priorities, plus formatting styles etc. 
 
> > I found the process quite long because we also improved several small things not related to the original idea, which was quite basic. I am also wondering how we could automatically integrate requested changes. 

Some of that is my own lack of knowledge in details of how we have flake8 configured. Apologies if this was painful for you.  

>Maintainers can push changes directly to the branch a PR was submitted from (minor >complication here is that @rlucas7 can't, since he doesn't yet have commit rights).

I'm not sure what additional responsibilities would be expected of me to have commit rights to the repo-or if that is even something the core devs would consider at this point. If it is, feel free to send me a message via my email to discuss.

>This is something where a reviewer has to find a balance. At the start it's good to point out all
>issues. 

Definitely still trying to find the right balance here, striving to be thorough while not turn off new committers. 
 
> Something that's harder to fix is that we're simply short on reviewer bandwidth. An all-volunteer > project with ~10 million users tends to be that way .....

@egouden here you could also consider reviewing as a way to contribute. As the contribution section of the repo docs state, reviewing code is also a very valuable contribution to the repo. 

> Let's see how your second PR goes:)

Yes! keep them coming.

**Comment by rlucas7** [comment]
> By the way, this PR broke our wheel tests because it raises a divide-by-zero warning sometimes: >https://travis-ci.org/MacPython/scipy-wheels/jobs/596997894. If you have time to look into it, >that would be very helpful. Otherwise I'll fix it later, but haven't had time yet.

@rgommers  it looks like the issue is because of the min diff in the (factored out) `_bin_numbers()` function. The smallest difference is 0 so the lines: 

```
690         try:                                                                    
691             decimal = int(-np.log10(dedges[i].min())) + 6                       
692         except RuntimeWarning:                                                  
693             raise ValueError('The smallest edge difference is numerically 0.') 
```

there are really 2 sources of issues in the calculation: 
1. the `dedges[i].min()` argument is 0 which causes a runtimeWarning to be thrown by the `np.log10()` call. 
2. the return value from `np.log10()` call is now `np.inf` and that cannot be converted to an `int` type by the call to `int()`

One solution could be to separate the line: 
`691             decimal = int(-np.log10(dedges[i].min())) + 6                       `
into 
```
log_dedges_min = -np.log10(dedges[i].min())
decimal = int(log_dedges_min)
```
so that if `dedges[i].min()==0.0` then the `np.log10()` call throws the runtime warning
separately from the `int()` typecast. 
currently the try/catch catches the runtimeWarning for the `dedges[i].min()==0.0` call but not the 
overflow error from call to `int(np.inf)`.

**Comment by rlucas7** [comment]
> One solution could be to separate the line:
> `691 decimal = int(-np.log10(dedges[i].min())) + 6 `

After looking into this a bit more, I'm not sure that is the right way to fix the problem. 
It looks like several places using the `numpy.errstate` context manager which might help in this scenario. I'll admit it is difficult to reproduce the failure in the travis build on my local machine.

I can open a PR if you want @rgommers but it is difficult for me to tell if that will fix the wheel build.

**Comment by egouden** [comment]
I just pushed a new version based on the comments.
What should I do now?

**Comment by rlucas7** [comment]
@egouden I think this is something Ralf and the core dev team will need to do on their end.

**Comment by rgommers** [comment]
> I can open a PR if you want @rgommers but it is difficult for me to tell if that will fix the wheel build.

that would be nice (unless @egouden already has the fix, see below)

> I pushed a new version based on the comments.
> What should I do now?
> Why are these additional tests not integrated in the PR tests?

Once your PR is merged, that's it - no changes will show up anymore in the GitHub UI, you can only open a new PR. You can do that from the same branch, but better would be to rebase your branch on current `master`. That way you only have the new commits that aren't merged yet in your branch.

**Comment by rgommers** [comment]
> @rgommers I'm not sure what additional responsibilities would be expected of me to have commit rights to the repo-or if that is even something the core devs would consider at this point. If it is, feel free to send me a message via my email to discuss.

I'll send you a message, let's chat:)

**Comment by rlucas7** [comment]
>that would be nice (unless @egouden already has the fix, see below)
>
>>   I pushed a new version based on the comments.
>>  What should I do now?
>>   Why are these additional tests not integrated in the PR tests?

I'm not sure if will fix the issue but I can open a PR tonight when I get home.

**Comment by rgommers** [comment]
@egouden it's not immediately obvious what your proposed fix is, looks like it's not a separate commit?

**Comment by egouden** [comment]
@rgommers Should I make a new PR?

**Comment by rgommers** [comment]
yes please!

**Comment by rgommers** [comment]
bug to reference for this new PR: https://github.com/MacPython/scipy-wheels/issues/53

(fyi @tylerjereddy)

**Pull request merged**
